src/lib/plugin_apis/crypto.api

```diff
@@ -1111,6 +1112,193 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
+    if (params == NULL)
+        return;
+
+    g_free (params->cipher);
+    g_free (params->cipher_mode);
+    g_free (params->resilience);
+    g_free (params->hash);
+    bd_crypto_luks_pbkdf_free(params->pbkdf);
+}
+
+/**
+ * bd_crypto_luks_reencrypt_params_new: (constructor)
+ * @key_size                    new volume key size if @new_volume_key is true. Ignored otherwise
+ * @cipher: (nullable):         new cipher
+ * @cipher_mode: (nullable):    new cipher mode
+ * @resilience: (nullable):     resilience mode to be used during reencryption
+ * @hash: (nullable):           used hash for "checksum" resilience type, ignored otherwise
+ * @max_hotzone_size            max hotzone size
+ * @sector_size                 sector size. Note that 0 is not a valid value
+ * @new_volume_key              whether to generate a new volume key or keep the existing one
+ * @offline                     whether to perform an offline or online reencryption,
+ *                                  i.e. whether a device is active in the time of reencryption or not
+ * @pbkdf: (nullable):          PBDKF function parameters for a new keyslot
+ */
+BDCryptoLUKSReencryptParams* bd_crypto_luks_reencrypt_params_new (guint32 key_size, gchar *cipher, gchar *cipher_mode, gchar *resilience, gchar *hash, guint64 max_hotzone_size, guint32 sector_size, gboolean new_volume_key, gboolean offline, BDCryptoLUKSPBKDF *pbkdf) {
+    BDCryptoLUKSReencryptParams *ret = g_new0 (BDCryptoLUKSReencryptParams, 1);
+    ret->key_size = key_size;
+    ret->cipher = g_strdup (cipher);
+    ret->cipher_mode = g_strdup (cipher_mode);
+    ret->resilience = g_strdup (resilience);
+    ret->hash = g_strdup (hash);
+    ret->max_hotzone_size = max_hotzone_size;
+    ret->sector_size = sector_size;
+    ret->new_volume_key = new_volume_key;
+    ret->offline = offline;
+    ret->pbkdf = bd_crypto_luks_pbkdf_copy(pbkdf);
+
+    return ret;
+}
+
+GType bd_crypto_luks_reencrypt_params_get_type () {
+    static GType type = 0;
+
+    if (G_UNLIKELY(type == 0)) {
+        type = g_boxed_type_register_static("BDCryptoLUKSReencryptParams",
+                                            (GBoxedCopyFunc) bd_crypto_luks_reencrypt_params_copy,
+                                            (GBoxedFreeFunc) bd_crypto_luks_reencrypt_params_free);
+    }
+
+    return type;
+}
+
+/**
+ * BDCryptoLUKSReencryptProgFunc:
+ * @size    size of the device being reencrypted
+ * @offset  current offset
+ *
+ * A callback function called during reencryption to report progress. Also used to possibly stop reencryption.
+ *
+ * Returns: 0, if the reencryption should continue.
+ *          A non-zero value to stop the reencryption
+ */
+typedef int (*BDCryptoLUKSReencryptProgFunc) (guint64 size, guint64 offset);
+
+typedef enum {
+    BD_CRYPTO_LUKS_REENCRYPT_NONE = 0,
+    BD_CRYPTO_LUKS_REENCRYPT_CLEAN,
+    BD_CRYPTO_LUKS_REENCRYPT_CRASH,
+    BD_CRYPTO_LUKS_REENCRYPT_INVALID
+} BDCryptoLUKSReencryptStatus;
```

Comment:
Looking at `bd_crypto_luks_reencrypt_status()`, does it make sense to include status values like "running" or "finished"?